qiskit/algorithms/optimizers/spsa.py

```diff
@@ -49,195 +53,335 @@ class SPSA(Optimizer):
+        self._nfev += 2
+
+        return gradient_sample
+
+    def _point_estimate(self, loss, x, eps, deltas):
+        """The gradient estimate at point ``x`` consisting as average of all directions ``delta``.
+        """
+        # number of samples
+        resamplings = len(deltas)
+
+        # set up variables to store averages
+        gradient_estimate = np.zeros(x.size)
+
+        # iterate over the directions
+        for delta in deltas:
+            gradient_sample = self._point_sample(loss, x, eps, delta)
+            gradient_estimate += gradient_sample
+
+        return gradient_estimate / resamplings
+
+    def _compute_update(self, loss, x, k, eps):
+        # compute the perturbations
+        if isinstance(self.resamplings, dict):
+            avg = self.resamplings.get(k, 1)
+        else:
+            avg = self.resamplings
+
+        gradient = np.zeros(x.size)
+
+        # accumulate the number of samples
+        deltas = [bernoulli_perturbation(x.size, self.perturbation_dims) for _ in range(avg)]
+
+        gradient = self._point_estimate(loss, x, eps, deltas)
+
+        return gradient
+
+    def _minimize(self, loss, initial_point):
+        # ensure learning rate and perturbation are set
+        # this happens only here because for the calibration the loss function is required
+        if self.learning_rate is None and self.perturbation is None:
+            get_learning_rate, get_perturbation = self.calibrate(loss, initial_point)
+            self.learning_rate = get_learning_rate
+            self.perturbation = get_perturbation
+
+        if self.learning_rate is None or self.perturbation is None:
+            raise ValueError('If one of learning rate or perturbation is set, both must be set.')
+
+        # get iterator
+        eta = self.learning_rate()
+        eps = self.perturbation()
+
+        # prepare some initials
+        x = np.asarray(initial_point)
+
+        self._nfev = 0
+
+        # if blocking is enabled we need to keep track of the function values
+        if self.blocking:
+            fx = loss(x)
+
+            self._nfev += 1
+            if self.allowed_increase is None:
+                self.allowed_increase = 2 * self.estimate_stddev(loss, x)
+
+        logger.info('=' * 30)
+        logger.info('Starting SPSA optimization')
+        start = time()
+
+        # keep track of the last few steps to return their average
+        last_steps = deque([x])
+
+        for k in range(1, self.maxiter + 1):
+            iteration_start = time()
+            # compute update
+            update = self._compute_update(loss, x, k, next(eps))
+
+            # trust region
+            if self.trust_region:
+                norm = np.linalg.norm(update)
+                if norm > 1:  # stop from dividing by 0
+                    update = update / norm
+
+            # compute next parameter value
+            update = update * next(eta)
+            x_next = x - update
```

Comment:
Do we not want to bound the variables in the range [0, 2pi] given that they are phase angles?

Reply:
I don't think we want to do this generally, since 
1. the angles can have coefficients 
2. it would imply you can only ever use SPSA for circuits (maybe you have a more generic loss function)
3. even if you only have circuits with coeffs equals to 1, it might be cheaper to "jump over" the 2pi boundary than "walk back", if that makes sense 

😄

Reply:
Not sure what a angle coefficient is.  I was thinking of a limited usage model where all parameters correspond to 1Q gate angles, and thus are cyclic mod 2pi.  But perhaps there is another use case for quantum that I am missing?  I am thinking that the routine here is not for general problems, if one desires a generic solver then why bury it here and not expose it as a stand-alone package or try adding it to SciPy as a general purpose optimizer?

Reply:
I mean that e.g this circuit is not cyclic mod 2 in `x` 🙂 
```
     ┌──────────┐
q_0: ┤ RY(x/10) ├
     └──────────┘
```